### PR TITLE
Fake db data

### PR DESCRIPTION
**Pull request by VirgoFrank**
## Fake db data
Add feature that creates fake data for database #117 

## Description

Add testing module
Requires faker package to work

```
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
```

### CODE REVIEW

**Comment by Orzelius** [comment]
Let's discuss this after todays standup

File: be/schema.gql [hunk 1]
```diff
@@ -49,6 +49,11 @@ type User {
   personal_code: String!
 }
 
+type Testing {
```
Comment (Orzelius) [inline be/schema.gql:52]: Update the schema

File: be/src/testing/testing.module.ts [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+import { TestingResolver } from './testing.resolver';
+import { ItemModule } from 'src/item/item.module';
+import { CategoryModule } from 'src/category/category.module';
+import { NoteModule } from 'src/note/note.module';
+import { UserModule } from 'src/user/user.module';
+
+@Module({
+  imports: [ItemModule, CategoryModule, NoteModule, UserModule],
+  providers: [TestingResolver, TestingService],
+  exports: [TestingService],
```
Comment (Orzelius) [inline be/src/testing/testing.module.ts:12]: why is this exported?

File: be/src/testing/testing.service.ts [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+@Injectable()
+export class TestingService {
+  constructor(
+    private readonly categoryRepo: CategoryService,
+    private readonly itemRepository: ItemService,
+    private readonly noteRepo: NoteService,
+    private readonly userRepo: UserService,
+  ) {
+    if (process.env.ADD_FAKE_DATA === 'TRUE') this.AddFakeData();
```
Comment (Orzelius) [inline be/src/testing/testing.service.ts:21]: remove this, let's gen fake data only when requested via gql api

File: be/src/testing/testing.service.ts [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+      throw new MethodNotAllowedException(`ADD_FAKE_DATA not set to TRUE`);
+    }
+    for (let i = 0; i < 3; i++) {
+      const item: CreateItemInput = { regCode: faker.datatype.number() };
+      this.itemRepository.createItem(item);
+
+      const category: CreateCategoryInput = {
+        name: faker.name.findName(),
+        bundelable: faker.datatype.boolean(),
+        parentCategoryId: null,
```
Comment (Orzelius) [inline be/src/testing/testing.service.ts:36]: maybe create a few child categories as well?

File: be/src/testing/testing.service.ts [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+      };
+      this.categoryRepo.create(category);
+
+      const note: CreateNoteInput = { content: faker.lorem.text(), itemId: 1 };
```
Comment (Orzelius) [inline be/src/testing/testing.service.ts:40]: Also add some notes to other items as well?

File: be/src/testing/testing.service.ts [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+      this.noteRepo.createNote(note);
+
+      const user: CreateUserDTO = {
+        username: faker.internet.userName(),
+        password: faker.internet.password(),
+        email: faker.internet.email(),
+        phone: faker.phone.phoneNumber(),
+        full_name: faker.name.findName(),
+        role: Math.floor(Math.random() * 2) === 1 ? Role.Admin : Role.User,
```
Comment (Orzelius) [inline be/src/testing/testing.service.ts:49]: The likelihood of this `=== 1` of being true is extremely small

File: be/src/testing/testing.service.ts [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+        registration_date: faker.datatype.datetime(),
```
Comment (Orzelius) [inline be/src/testing/testing.service.ts:50]: Could this possible give a future date?

File: be/src/testing/testing.service.ts [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+    private readonly categoryRepo: CategoryService,
+    private readonly itemRepository: ItemService,
+    private readonly noteRepo: NoteService,
+    private readonly userRepo: UserService,
+  ) {
+    if (process.env.ADD_FAKE_DATA === 'TRUE') this.AddFakeData();
+    console.log('add fake data ', process.env.ADD_FAKE_DATA);
+  }
+
+  AddFakeData() {
```
Comment (Orzelius) [inline be/src/testing/testing.service.ts:25]: I think overall this function could be a bit more advanced, but I guess for now this will do

**Review by Orzelius — CHANGES_REQUESTED** [review]

**Comment by Orzelius** [comment]
And you still have some lint errors

File: be/src/testing/testing.service.ts (hunk 6 above)
Comment (VirgoFrank) [reply be/src/testing/testing.service.ts:49]: No, Math.random() * 2 generates a number between 0 and 2 then Math.floor method rounds it down to an integer. This generates either 0 or 1.

Comment (Orzelius) [reply be/src/testing/testing.service.ts:49]: ah, yep. I didn't notice the `Math.floor`, my bad

File: be/src/testing/testing.service.ts (hunk 5 above)
Comment (VirgoFrank) [reply be/src/testing/testing.service.ts:40]: I don't really know how i can easily add other ids. Should i get the number of ids from the db and then generate a number between the number of ids and 1? Or would there be some other way?

Comment (Orzelius) [reply be/src/testing/testing.service.ts:40]: The `createItem` function you are using returns the item it had created. You can then store the id's of the items and add notes to them

File: be/src/testing/testing.service.ts (hunk 7 above)
Comment (VirgoFrank) [reply be/src/testing/testing.service.ts:50]: Yes. I checked I could use date.recent() to give a date that is one or two days ago.

**Comment by Orzelius** [comment]
@VirgoFrank When you feel like you're ready for second round of review just request a new review
image

File: be/schema.gql (hunk 1 above)
Comment (Orzelius) [reply be/schema.gql]: done in -> 6d8d167

File: be/src/testing/testing.service.ts (hunk 7 above)
Comment (Orzelius) [reply be/src/testing/testing.service.ts:58]: I think in this case it doesn't make sense to have a future date in the `registration_date` field

File: be/src/testing/testing.service.ts (hunk 5 above)
Comment (VirgoFrank) [reply be/src/testing/testing.service.ts]: done in -> [98ab51](https://github.com/TPT-Loane/TPT-Loane/pull/120/commits/98ab5169b3b4faf4959945cbe16e8b67b06e2ed4)

File: be/src/testing/testing.service.ts (hunk 4 above)
Comment (VirgoFrank) [reply be/src/testing/testing.service.ts:33]: done in [a35339](https://github.com/TPT-Loane/TPT-Loane/pull/120/commits/a353392cbcddaaad24948aba9a4e01545c95f900) and [4e2c5b](https://github.com/TPT-Loane/TPT-Loane/pull/120/commits/4e2c5b4cf733f655831cd982cdfd397411a33e1d)

File: be/src/testing/testing.service.ts (hunk 3 above)
Comment (VirgoFrank) [reply be/src/testing/testing.service.ts]: done in [9f7d52](https://github.com/TPT-Loane/TPT-Loane/pull/120/commits/9f7d524ac3c0650a0d2d8f89b5df19a798028542)

File: be/src/testing/testing.module.ts (hunk 2 above)
Comment (VirgoFrank) [reply be/src/testing/testing.module.ts]: removed in [0095b0](https://github.com/TPT-Loane/TPT-Loane/pull/120/commits/0095b0db85252a7760b0493ae5789556f1ba628e)

**Review by Orzelius — APPROVED** [review]
If you can be bothered you could fix the `registration_date` issue, but otherwise all good 👍

File: be/src/testing/testing.service.ts (hunk 7 above)
Comment (VirgoFrank) [reply be/src/testing/testing.service.ts]: Changed it to date.recent() here [fbcc74a](https://github.com/TPT-Loane/TPT-Loane/pull/120/commits/fbcc74a768a6ba3261dcf386ea58aa65c5b2df66)

**Review by Orzelius — APPROVED** [review]
👍

**Review by Erikyy — DISMISSED** [review]

**Pull request merged**